src/sql/src/plan/statement/dml.rs

```diff
@@ -457,16 +485,85 @@ pub fn plan_subscribe(
                 QueryLifetime::OneShot(scx.pcx()?),
             )?;
             assert!(query.finishing.is_trivial(query.desc.arity()));
-            SubscribeFrom::Query {
-                expr: query.expr,
-                desc: query.desc,
-            }
+            let desc = query.desc.clone();
+            (
+                SubscribeFrom::Query {
+                    expr: query.expr,
+                    desc: query.desc,
+                },
+                desc,
+                query.scope,
+            )
         }
     };
 
     let when = query::plan_as_of(scx, as_of)?;
     let up_to = up_to.map(|up_to| plan_up_to(scx, up_to)).transpose()?;
 
+    let qcx = QueryContext::root(scx, QueryLifetime::OneShot(scx.pcx()?));
+    let ecx = ExprContext {
+        qcx: &qcx,
+        name: "",
+        scope: &scope,
+        relation_type: desc.typ(),
+        allow_aggregates: false,
+        allow_subqueries: true,
+        allow_windows: false,
+    };
+
+    let output = match output {
+        SubscribeOutput::Diffs => plan::SubscribeOutput::Diffs,
+        SubscribeOutput::EnvelopeUpsert { key_columns } => {
+            scx.require_envelope_upsert_in_subscribe()?;
+            let order_by = key_columns
+                .iter()
+                .map(|ident| OrderByExpr {
+                    expr: Expr::Identifier(vec![ident.clone()]),
+                    asc: None,
+                    nulls_last: None,
+                })
+                .collect_vec();
+            let (order_by, map_exprs) = query::plan_order_by_exprs(
+                &ExprContext {
+                    name: "ENVELOPE UPSERT KEY clause",
+                    ..ecx
+                },
+                &order_by[..],
+                &[],
+            )?;
+            if !map_exprs.is_empty() {
+                return Err(PlanError::InvalidKeysInSubscribeEnvelopeUpsert);
+            }
+            plan::SubscribeOutput::EnvelopeUpsert {
+                order_by_keys: order_by,
+            }
+        }
+        SubscribeOutput::WithinTimestampOrderBy { order_by } => {
+            scx.require_within_timestamp_order_by_in_subscribe()?;
+            let mz_diff_fake_column = usize::MAX;
+            let (mut order_by, map_exprs) = query::plan_order_by_exprs(
```

Comment:
There is something fishy going on in here:

```
DROP TABLE IF EXISTS t1 CASCADE;
CREATE TABLE t1 (f1 INTEGER, f2 INTEGER);
COPY (SUBSCRIBE t1 WITH (SNAPSHOT=TRUE) WITHIN TIMESTAMP ORDER BY 2) TO STDOUT;
ERROR:  column reference 2 in WITHIN TIMESTAMP ORDER BY clause is out of range (1 - 1)

```

In this context, `2` means order by the second column of the result set, and indeed this works with a stand-alone `SELECT` statement against the same table. Is the `mz_diff` column already available in this code block? If `ORDER BY N` is ambiguous or otherwise problematic it may be best to disallow it altogether in this context.

Reply:
good catch, thanks!